### DISCUSSION TITLE
Adding support for IP location data (geoIP)

### DISCUSSION THREAD

**Discussion by snebel29** [coredns/coredns#4685] [General]
Hi,
I've been thinking about adding support for geoIP and I would like to discuss this possibility, it's my understanding that such functionality would be welcomed based on https://github.com/coredns/coredns/issues/513 and https://github.com/coredns/coredns/issues/143.

Assuming that there is no push back on the idea, I was planning to create an internal plugin (disabled by default) that would add support for looking up [maxmind geoIP2 databases](https://www.maxmind.com/en/geoip2-databases) that seems to be the facto IP databases out there providing geo IP database but not limited to, I was thinking that this plugin could be consumed in two modes.

1. **By hooking it up to the chain of plugins**: Where it would automatically would read the client source IP, lookup in the configured databases and add the data associated with that IP as a request context value, other plugins could find out if the plugin was enabled to during setup/initialization, then lookup directly from the context. This mode would be useful in case multiple plugins want to make use of the same database data in a single request, may also useful if other plugins want to enrich the data.
2. **By wrapping and exposing geoIP methods**: In some cases, plugins looking up in the databases would be the only consumers and may want to import the plugin and use directly.

After this plugin to lookup in maxmind databases is ready to use, adding support for closest geo IP in existings plugins such as file, auto, etc. would be relatively easy.

How does that sound? I volunteer to create the pull request, and may provide a draft if that helps to picture the design.

Thanks in advance.

**Comment by miekg** [comment]
[ Quoting ***@***.***> in "[coredns/coredns] Adding support fo..." ]
> 1. By hooking it up to the chain of plugins: Where it would automatically
>    would read the client source IP, lookup in the configured databases and add
>    the data associated with that IP as a request context value, other plugins
>    could find out if the plugin was enabled to during setup/initialization,
>    then lookup directly from the context. This mode would be useful in case
>    multiple plugins want to make use of the same database data in a single
>    request, may also useful if other plugins want to enrich the data.

yes, the metadata plugin/code can be used for this.

> 2. By wrapping and exposing geoIP methods: In some cases, plugins looking up
>    in the databases would be the only consumers and may want to import the
>    plugin and use directly.

that may not even be needed if the metadata approach gives enough info.

>After this plugin to lookup in maxmind databases is ready to use, adding
>support for closest geo IP in existings plugins such as file, auto, etc. would
>be relatively easy.
>
>How does that sound?

sgtm, but do keep in mind that any external lookup can't reliably use DNS names as we
can't guarantee a working DNS env. I.e. get the maxmind db should be an offline process
outside of coredns or you need to connect to it via IP addresses.

/Miek

--
Miek Gieben

**Comment by snebel29** [comment]
Hi,
Thanks for your quick answer

> yes, the metadata plugin/code can be used for this.

Great, I didn't notice it existed. I will certainly look to it and use it.

>  that may not even be needed if the metadata approach gives enough info.

Ok, I was mildly feeling this would be easy to just export the right methods in case someone has a use case and decide to import instead, but I guess whether this is appropriate can be better discussed inside the upcoming PR.

> sgtm, but do keep in mind that any external lookup can't reliably use DNS names as we
can't guarantee a working DNS env. I.e. get the maxmind db should be an offline process
outside of coredns or you need to connect to it via IP addresses.

Yes, so the idea would be to pass the database path which has been previously downloaded in a plugin config block, then use it for opening. The plugin would fail (or may be just ignore the plugin with a warning) if it doesn't manage to open the configured database, no online download of the database, or extra DNS lookup of any kind.

My net steps:

- Take a look to last plugin PRs to get a sense of the size of PR and how are they typically break down (multiple small PR, vs one large where we add different phases in commits then squash for example) to make sure it's easy to review, comment, etc. But if you have a preference just let me know.
-  I may create something, may me just to show the structure with comments and TODOs to get early feedback and move forward from there.

Thanks in advance.

**Comment by miekg** [comment]
[ Quoting ***@***.***> in "Re: [coredns/coredns] Adding suppor..." ]
>My net steps:
>
>  • Take a look to last plugin PRs to get a sense of the size of PR and how are
>    they typically break down (multiple small PR, vs one large where we add
>    different phases in commits then squash for example) to make sure it's easy
>    to review, comment, etc. But if you have a preference just let me know.
>  • I may create something, may me just to show the structure with comments and
>    TODOs to get early feedback and move forward from there.

ack. Note I guess this will be a rather small plugin, as it's mostly setting some meta
data.

Usually a new plugin starts by commiting the README.md.

**Comment by snebel29** [comment]
Hi,
I took a look to the metadata plugin, it's a great fit for what we are discussing here, I see one "challenge" in using it, the function signature returned by `ValueFun` is `func() string` which is good for most use cases.
 
However in the case of max mind databases (geoIP2), the data record associated with an IP is a struct with multiple fields and different types, see example for City database [here](https://github.com/oschwald/geoip2-golang/blob/f8b05833099d49d922c5e5b88791375f03e6628b/reader.go#L85-L133). Ideally we would want to provide other plugins/middlewares access to all the associated information.

> Note: Users may even create their own maxminddb database and associating data with arbitrary structures.

Several solutions for mmdb/geoip2 plugin to be able to use the metadata plugin comes to my mind, there may be others though:

### 1. Leverage DatabaseType field.
Every maxminddb comes with a field [DatabaseType](https://github.com/oschwald/maxminddb-golang/blob/55ce382ad3ce5b08b6a70e9448a33815d232d896/reader.go#L46) string field in its own metadata database section, this is what `oschwald/geoip2-golang` reader package is already doing to figure out the database and decode the results into the right structs. 

We could store that field in metadata plugin, then metadata consumers could use it for doing an extra lookup to some exported method of the new `mmdb/geoip2` plugin and retrieve the desired struct, for official databases and an empty interface for custom ones, that consumers would ultimately have to cast.

**Pros**: No need to change metadata plugin
**Cons**: It's somehow defeating the role of metadata plugin, to do this way why not to completely work it around?

### 2. Serialize the data
At first glance this feels like an expensive approach for both the producer and the consumer.

### 3. Extend the metadata plugin to allow `func() interface{}` functions.
The title is self explanatory, this would allow consumers to cast the value based on the key label, example: `plugin/mmdb-city` where city matches databaseType `city`, would return an empty interface that can be casted to a city struct.

**Pros**: Works with both official and custom mmdb databases, allows passing context key values != string in metadata plugin for other potential future plugins, should be fast.
**Cons**: Beyond having to add further functions to metadata plugin I see none.

### 4. Flatten individual values into the context
The `mmdb/geoip2` plugin would flatten the values in the struct to individual key values in the context, for example:

`plugin/mmdb-iso-code` -> 123456
`plugin/mmdb-city-name` -> Amsterdam
...

**Pros**: No need to change metadata plugin, easy for users to consume the key values (no casting, etc.)
**Cons**: You still may have to serialize types than are composed of other types, or to represent them in complex keys making it hard to read and find out, for example having to suffix keys with indexed for slices, etc.

**I'm feeling number 3 is overall a better solution**, what do you think? do you have any other idea or suggestion?

Sorry in advance for the long text,

Cheers

**Reply by johnbelamaric** [reply]
The problem with number 3 is that the consuming plugins need to all understand the types, which creates more dependencies between plugins. For this reason I think number 4 works better. It minimizes inter-plugin dependencies at the code level (there is still dependency at the specific key definitions, of course).

**Reply by johnbelamaric** [reply]
A middle-ground would be allowing the valueFunc to return a map of strings. But only if it can be done in a way that requires no updates to existing plugins.

**Comment by miekg** [comment]
Multiple metadata labels and values are fine with me. Suggestion for the
syntax is to use another / to separate the sub labels

**Comment by snebel29** [comment]
> The problem with number 3 is that the consuming plugins need to all understand the types, which creates more dependencies between plugins. For this reason I think number 4 works better. It minimizes inter-plugin dependencies at the code level (there is still dependency at the specific key definitions, of course).

**Thinking out loudly**: I agree, however number 3 feels like the only reasonable way (without using reflect) to allow the use of arbitrary database structures in the future, other small (may be even negligible) downside of number 4 is the cost of having to convert numeric datatypes to string and then back to numbers in other plugins, for example for longitude and latitude fields which I would expect to be one of the most common use cases.

May be the answer is to not support (at least not for now) custom databases with arbitrary structures and focus in the official databases with known schemas/structures which are the most likely common use cases, and go with multiple label/values (option 4), which you and @miekg  seems to be happy with.

> Multiple metadata labels and values are fine with me. Suggestion for the
syntax is to use another / to separate the sub labels

Sounds good, I'll do that.

I'll give it a try during the weekend, and share.

Thank you all for your time.
